c++/src/ColumnReader.hh

```diff
@@ -58,6 +58,21 @@ namespace orc {
                               proto::Stream_Kind kind,
                               bool shouldStream) const = 0;
 
+    std::unique_ptr<SeekableInputStream>
+                    getStreamOrThrow(const Type& type,
```

Comment:
I prefer to make it virtual and remove getStream interface.

Reply:
I did not want to remove it since I saw it used here without throwing error:
https://github.com/apache/orc/blob/ffb1a2b51f25b909a1d5da1090f8b471b8fca286/c%2B%2B/src/ColumnReader.cc#L54
Should we throw there or not?

How about making getStream protected while keeping getStreamOrThrow non-virtual?

Reply:
Definitely not throw there. How about adding a default input parameter in getStream to advise whether it should throw or not if stream is null?
` getStream(uint64_t columnId, proto::Stream_Kind kind, bool shouldStream, bool optional = false)`

Reply:
Actually proto::Stream_Kind is sufficient to provide whether it is optional or not.

Reply:
Okay, so if I understand correctly here's how it should look like:
1. The interface should not change (or would only change in replacing columnId with type in order to give better debug information).
2. The check should happen in getStream implementation.
3. If the stream kind is not PRESENT, we should throw an exception (maybe through a private/protected function to get standard error messages).

What do you think?

Reply:
I agree on most of them. For 3), I think it is better to pass `bool optional` explicitly. The reason is that not only PRESENT stream can be optional but also SECONDARY stream of decimal type in ORC 2.0.